src/Compilers/Core/Portable/Text/StringTextWriter.cs

```diff
@@ -2,29 +2,29 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
-using System.Collections.Generic;
-using System.Collections.Immutable;
-using System.Diagnostics;
-using System.IO;
-using System.Linq;
 using System.Text;
-using System.Threading;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Text
 {
-    internal class StringTextWriter : SourceTextWriter
+    internal sealed class StringTextWriter : SourceTextWriter
     {
         private readonly StringBuilder _builder;
         private readonly Encoding? _encoding;
         private readonly SourceHashAlgorithm _checksumAlgorithm;
 
-        public StringTextWriter(Encoding? encoding, SourceHashAlgorithm checksumAlgorithm, int capacity)
+#if DEBUG
+        private readonly int _length;
+#endif
+
+        public StringTextWriter(Encoding? encoding, SourceHashAlgorithm checksumAlgorithm, int length)
         {
-            _builder = new StringBuilder(capacity);
+            _builder = new StringBuilder(length);
```

Comment:
🤔 Is there a reason why this isn't using `PooledStringBuilder`? It seems like all other places in the compiler that use a `StringBuilder` use the pooled one.

Reply:
size is likely so large that pooling would be undesirable.  Also, this scales with # files, which is low.  We want pooling for when tehre are literally insane numbers of allocations otherwise.

Reply:
`PooledStringBuilder` already has a check for the size and only pools ones smaller than 1024K chars.

Reply:
Given that `SourceTextWriter` is used inside `CompositeText`, which is used by `WithChanges`, I'd say the lengths here probably are very small most often.

Reply:
I made a separate PR for this: #69386